registry/federated-namespaces.md

```diff
@@ -0,0 +1,155 @@
+   1. `registry`: The hostname of the registry with which to interact.
+   1. `namespace`: The namespace (on DockerHub the user or organization) under which the image exists.
+   1. `image:tag`: This is further split on `:` into at most 2 pieces, with `tag` defaulting to `latest`.
+
+These components are validated against the regular expressions (see [validateRemoteName](https://sourcegraph.com/github.com/docker/docker@624de8a9cd5652d0d685596d490997f7f2e83536/.tree/registry/config.go#startline=205&endline=205), see [ValidateTagName](https://sourcegraph.com/github.com/docker/docker@624de8a9cd5652d0d685596d490997f7f2e83536/.tree/graph/tags.go#startline=354&endline=354)):
+   1. `namespace` must match `^([a-z0-9-_]*)$`
+   1. `image` must match `^([a-z0-9-_.]+)$`
+   1. `tag` must match `^[\w][\w.-]{0,127}$`
+
+### v2
+
+In the world of Docker Registry v2, names have a similar, but slightly more flexible form best described [here](https://sourcegraph.com/github.com/docker/distribution@5556cd1ba11aea6803f3c51a16bb96813ba72800/.tree/registry/api/v2/names.go#selected=56):
+```
+// Effectively, the name should comply with the following grammar:
+//
+// 	alpha-numeric := /[a-z0-9]+/
+//	separator := /[._-]/
+//	component := alpha-numeric [separator alpha-numeric]*
+//	namespace := component ['/' component]*
+```
+Where `namespace` is used in the context: `registry/namespace:tag`, where as before `registry` and `tag` are optional.
+
+Processing this image path simplifies to:
+   1. If there are multiple parts, then `registry` is the first part if the first part is `localhost`, or contains a `:` or `.`.
+   1. Otherwise, `registry` is `index.docker.io`.
+
+The final component of the remainder is split on `:` into two parts, and the second half becomes the `tag`, which I believe to be validated in the same fashion as in v1.
+
+What remains is validated against `namespace` in the above BNF grammar, and limited to `255` characters.
+
+
+### v1 vs. v2
+
+The acceptable names in v1 vs. v2 form a Venn diagram.  This covers several examples.
+
+#### v1 only
+```
+registry/_foo/bar:baz
+registry/foo/a-name-longer-than-255-characters:baz
+registry/a-255-character-name/bar:baz
+```
+
+**NOTE**: The [Google Container Registry](https://gcr.io) makes extensive use of `gcr.io/_b_...` style prefixes to alter behaviors.
+
+#### v2 only
+```
+registry/f.o.o/bar:baz
+registry/f/o/o/bar:baz
+```
+
+#### both
+```
+registry/foo/bar:baz
+```
+
+## Part 1: Relax client-side validation
+
+In order to continue to support both v1 and v2 within the same client, while enabling customers to start leveraging the more flexible aspects of naming afforded by the v2 registry, I propose the following grammar for image paths:
+```
+//  tag := /[^:]+/
+//  hash := 'sha256'
+//        | 'sha284'
+//        | 'sha512'
+//        | 'tarsum'     // TODO(mattmoor): Can we retire this?
+//  digest := /[a-fA-F0-9]+/
+//  image := /.+/
```

Comment:
FWIW, by making the v2-naming space a non-superset of v1, you have already created this situation.  In fact, I have included an example where I do precisely this with the 1.6 client towards the end.

Moreover, you cannot possibly allow folks to *start* using the expanded namespace afforded by v2 without considering the case where it fails when pushing to v1.